File: src/main/java/at/searles/parsing/utils/common/ToString.java
package at.searles.parsing.utils.common;

import at.searles.parsing.ParserCallBack;
import at.searles.parsing.Mapping;
import at.searles.parsing.ParserStream;
import at.searles.parsing.PrinterCallBack;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class ToString implements Mapping<CharSequence, String> {

    public static ToString getInstance() {
        return Holder.INSTANCE;
    }

    @Override
    public String parse(ParserCallBack env, ParserStream stream, @NotNull CharSequence left) {
        return left.toString();
    }

    @Nullable
    @Override
    public CharSequence left(PrinterCallBack env, @NotNull String result) {
        return result;
    }

    private static class Holder {
        static final ToString INSTANCE = new ToString();
    }
}

File: src/main/java/at/searles/parsingtools/common/ToString.kt
package at.searles.parsingtools.common

import at.searles.parsing.Mapping
import at.searles.parsing.ParserStream

object ToString : Mapping<CharSequence, String> {
    override fun reduce(left: CharSequence, stream: ParserStream): String {
        return left.toString()
    }

    override fun left(result: String): CharSequence {
        return result
    }

    override fun toString(): String {
        return "{x -> x.toString}"
    }
}
